[CI]  MixedClusterClientYamlTestSuiteIT failing (crash due to assertion)

Has happened on two different PRs and locally on unmodified master

**Build scan**:
https://gradle-enterprise.elastic.co/s/3wwzm7vghnze4
https://gradle-enterprise.elastic.co/s/3jiec6et2wbbe

**Repro line**:

```
./gradlew ':qa:mixed-cluster:v7.15.0#mixedClusterTest' -Dtests.class="org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT" -Dtests.method="test {p0=snapshot.get/10_basic/Get snapshot info when verbose is false}" -Dtests.seed=489DEE34D256CCBA -Dtests.bwc=true -Dtests.locale=sl-SI -Dtests.timezone=America/Cayman -Druntime.java=11
```

**Reproduces locally?**:

Yes, using following against master:
```
./gradlew ':qa:mixed-cluster:check' -Dtests.seed=A5080EDD8FCF06C8 -Dtests.bwc=true -Dtests.locale=es-HN -Dtests.timezone=Etc/GMT+9 -Druntime.java=11
```

**Applicable branches**:

master (at least).

**Failure history**:

[build-stats](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-30d,mode:quick,to:now))&_a=(columns:!(_source),index:f0e86d70-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'%22Provided%20explict%20repo%20shard%20id%20statuses%22'),sort:!(time,desc)))

started failing this morning. Likely the backport of #75501 in  #76539.

**Failure excerpt**:

```
[2021-08-16T07:44:25,603][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [v7.15.0-0] fatal error in thread [elasticsearch[v7.15.0-0][generic][T#4]], exiting
java.lang.AssertionError: Provided explict repo shard id statuses [[RepositoryShardId{[test_index/84YPuuceTUKHo8NGh3X_Xw]}{0}=>ShardSnapshotStatus[state=INIT, nodeId=qccWoxkyR7-hl0j9W2SRlA, reason=null, generation=_new, shardSnapshotResult=null]]] but no source
	at org.elasticsearch.cluster.SnapshotsInProgress$Entry.<init>(SnapshotsInProgress.java:559) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.SnapshotsInProgress$Entry.readFrom(SnapshotsInProgress.java:610) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.SnapshotsInProgress.lambda$new$0(SnapshotsInProgress.java:67) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.io.stream.StreamInput.readCollection(StreamInput.java:1253) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.common.io.stream.StreamInput.readList(StreamInput.java:1203) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.cluster.SnapshotsInProgress.<init>(SnapshotsInProgress.java:67) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
```

Did not mute any tests, since this hits too widely.

--- a/server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
+++ b/server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java
@@ -558,7 +558,7 @@ public class SnapshotsInProgress extends AbstractNamedDiffable<Custom> implement
                       List<String> dataStreams, List<SnapshotFeatureInfo> featureStates, long startTime, long repositoryStateId,
                       ImmutableOpenMap<ShardId, ShardSnapshotStatus> shards, String failure, Map<String, Object> userMetadata,
                       Version version, @Nullable SnapshotId source,
-                      ImmutableOpenMap<RepositoryShardId, ShardSnapshotStatus> shardStatusByRepoShardId) {
+                      @Nullable ImmutableOpenMap<RepositoryShardId, ShardSnapshotStatus> shardStatusByRepoShardId) {
             this.state = state;
             this.snapshot = snapshot;
             this.includeGlobalState = includeGlobalState;
@@ -658,7 +658,7 @@ public class SnapshotsInProgress extends AbstractNamedDiffable<Custom> implement
             }
             return new SnapshotsInProgress.Entry(
                     snapshot, includeGlobalState, partial, state, indices, dataStreams, featureStates, startTime, repositoryStateId,
-                    shards, failure, userMetadata, version, source, source == null ? null : clones);
+                    shards, failure, userMetadata, version, source, clones);
         }
 
         private static boolean assertShardsConsistent(SnapshotId source, State state, Map<String, IndexId> indices,
@@ -1028,7 +1028,11 @@ public class SnapshotsInProgress extends AbstractNamedDiffable<Custom> implement
             }
             if (out.getVersion().onOrAfter(SnapshotsService.CLONE_SNAPSHOT_VERSION)) {
                 out.writeOptionalWriteable(source);
-                out.writeMap(shardStatusByRepoShardId);
+                if (source == null) {
+                    out.writeMap(ImmutableOpenMap.of());
+                } else {
+                    out.writeMap(shardStatusByRepoShardId);
+                }
             }
             if (out.getVersion().onOrAfter(FEATURE_STATES_VERSION)) {
                 out.writeList(featureStates);

